### PR TITLE
Adjust sds types

### PR DESCRIPTION
**Pull request by poiuj**
sds type should be determined based on the size of the underlying buffer, not the logical length of the sds. Currently we truncate the alloc field in case buffer is larger than we can handle. It leads to a mismatch between alloc field and the actual size of the buffer. Even considering that alloc doesn't include header size and the null terminator.

It also leads to a waste of memory with jemalloc. For example, let's consider creation of sds of length 253. According to the length, the appropriate type is SDS_TYPE_8. But we allocate `253 + sizeof(struct sdshdr8) + 1` bytes, which sums to 257 bytes. In this case jemalloc allocates buffer from the next size bucket. With current configuration on Linux it's 320 bytes. So we end up with 320 bytes buffer, while we can't address more than 255.

The same happens with other types and length close enough to the appropriate powers of 2.

The downside of the adjustment is that with allocators that do not allocate larger than requested chunks (like GNU allocator), we switch to a larger type "too early". It leads to small waste of memory. Specifically: sds of length 31 takes 35 bytes instead of 33 (2 bytes wasted) sds of length 255 takes 261 bytes instead of 259 (2 bytes wasted) sds of length 65,535 takes 65,545 bytes instead of 65,541 (4 bytes wasted) sds of length 4,294,967,295 takes 4,294,967,313 bytes instead of 4,294,967,305 (8 bytes wasted)

### CODE REVIEW

**Comment by poiuj** [comment]
This patch is subset of changes in https://github.com/valkey-io/valkey/pull/453. I think we can merge this independently. https://github.com/valkey-io/valkey/pull/453 depends on this patch and on https://github.com/valkey-io/valkey/pull/476.

File: src/sds.c [hunk 1]
```diff
@@ -124,8 +110,6 @@ sds _sdsnewlen(const void *init, size_t initlen, int trymalloc) {
     s = (char*)sh+hdrlen;
     fp = ((unsigned char*)s)-1;
     usable = usable-hdrlen-1;
-    if (usable > sdsTypeMaxSize(type))
```
Comment (madolson) [inline src/sds.c]: The defensive programmer in me wants to replace this with an assert that usable <= the max size.

File: src/sds.c [hunk 2]
```diff
@@ -59,35 +59,21 @@ static inline int sdsHdrSize(char type) {
 }
 
 static inline char sdsReqType(size_t string_size) {
-    if (string_size < 1<<5)
+    if (sizeof(struct sdshdr5) + string_size + 1 <= 1<<5)
         return SDS_TYPE_5;
-    if (string_size < 1<<8)
+    if (sizeof(struct sdshdr8) + string_size + 1 <= 1<<8)
```
Comment (madolson) [inline src/sds.c]: If possible, I'd like to document this better. The `<=` behavior was surprising to convince myself it was correct, because the actual value of ->alloc doesn't include the overhead, so we have 2-9 bytes of space which happens to always be enough room.

**Review by madolson — COMMENTED** [review]
I think this makes sense to me. The slightly more memory usage on some systems is likely OK since I think we still prefer to use jemalloc in most cases.

File: src/sds.c (hunk 2 above)
Comment (poiuj) [reply src/sds.c]: I agree. Let me try to improve the documentation here.

**Comment by ranshid** [comment]
@poiuj overall looks much better now. I would suggest we include some unit tests to cover different types edge cases (for example the case of old style alloc overflow). I am comfortable with the asserts, but we will probably want to trigger ci and daily before we add this.

After we finish this PR lets go and fix #453 accordingly

File: src/sds.c (hunk 1 above)
Comment (poiuj) [reply src/sds.c]: Done

**Comment by ranshid** [comment]
@poiuj the tests looks fine. 
Regarding the top comment:

>The downside of the adjustment is that with allocators that do not allocate larger than requested chunks (like GNU allocator), we switch to a larger type "too early". It leads to small waste of memory. Specifically: sds of length 31 takes 35 bytes instead of 33 (2 bytes wasted) sds of length 255 takes 261 bytes instead of 259 (2 bytes wasted) sds of length 65,535 takes 65,545 bytes instead of 65,541 (4 bytes wasted) sds of length 4,294,967,295 takes 4,294,967,313 bytes instead of 4,294,967,305 (8 bytes wasted)

Is it really a waste? we already reduced the overflow in the prev implementation and AFAIK we never should have written pass the usable anyhow right? 
If anything this would have solved some fragmentation problem IIUC:

lets say I will allocate 253 sds.

Before: I would allocate according to 253+(sizeof(sdshdr8)=3)+1 =  257 --> allocated size of 320 (IIRC) but the usable size is 255 
After:  I would allocate according to 253+(sizeof(sdshdr16)=5)+1 = 259 --> allocated size of 320 (IIRC) and the usable size is 314 

So IMO this is 59 bytes gain:) 
I think the main gain here is that we will be able to avoid mallocsize in realloc (and the optimization in #453)

I still want to verify there are no issues related to backward compatibility (I think RDB does not serialize sds headers, but would like to think on other potential case)

**Comment by poiuj** [comment]
@ranshid your example is correct. It's a 59 bytes gain if we use jemalloc. My top commented is about libc allocator that wouldn't allocate 320 bytes in this case.

**Comment by ranshid** [comment]
> @ranshid your example is correct. It's a 59 bytes gain if we use jemalloc. My top commented is about libc allocator that wouldn't allocate 320 bytes in this case.

Oh I get it now. yes. indeed this is not too bad IMO. Please also consider avoid the malloc_size use in realloc

**Comment by poiuj** [comment]
@ranshid sure, I plan to get rid of call to `zmalloc_size` in resize as part of #453.

File: src/sds.c [hunk 3]
```diff
@@ -58,36 +58,42 @@ static inline int sdsHdrSize(char type) {
     return 0;
 }
 
+/* Returns the maximum len and alloc for a given type.
+ *
+ * The buffer includes header and null terminator. The total buffer size must be
+ * up to 2^5, 2^8, 2^16, and 2^32 bytes. We assume jemalloc uses these sizes for
+ * allocation. Otherwise jemalloc allocates larger buffer, which length can't be
+ * stored in alloc field. */
+static inline size_t sdsTypeMaxSize(char type) {
```
Comment (madolson) [inline src/sds.c]: Smallest of comments, can we move this function back to where it was so we keep the git history.

File: src/sds.c [hunk 4]
```diff
@@ -124,8 +130,7 @@ sds _sdsnewlen(const void *init, size_t initlen, int trymalloc) {
     s = (char*)sh+hdrlen;
     fp = ((unsigned char*)s)-1;
     usable = usable-hdrlen-1;
-    if (usable > sdsTypeMaxSize(type))
-        usable = sdsTypeMaxSize(type);
+    assert(usable <= sdsTypeMaxSize(type));
```
Comment (zhulipeng) [inline src/sds.c]: You can't do such assertion here, usable is got from allocator and allocator didn't have such kind of guarantee that it will have maximum size. For example, if I requested 236 bytes, the allocator size could be 248 bytes (good) but maybe 264 bytes. I think the guarantee is that the allocated size >= requested size.

File: src/sds.c (hunk 3 above)
Comment (poiuj) [reply src/sds.c]: We now use it in `sdsReqType`. Do you prefer to add a forward declaration of `sdsTypeMaxSize` to keep the git history?

File: src/sds.c (hunk 4 above)
Comment (poiuj) [reply src/sds.c]: You're right, the allocator doesn't guarantee a maximum size. But it doesn't give random blocks either. Jemalloc allocates buffers of pre-defined sizes. And SDS types are aligned accordingly.
If this assertion is false, `usable` is incorrect.

Comment (poiuj) [reply src/sds.c]: I thought about jemalloc. I was sure that libc allocator would return a buffer of the exact size. Apparently I was wrong, thank you for noting it. I'll fix it.

File: src/sds.c (hunk 3 above)
Comment (madolson) [reply src/sds.c]: yes

**Review by madolson — APPROVED** [review]
LGTM, but would like @lipzhu and @ranshid to signoff as well. This is sensitive code and would like extra eyes on it.

File: src/sds.c [hunk 5]
```diff
@@ -124,8 +132,14 @@ sds _sdsnewlen(const void *init, size_t initlen, int trymalloc) {
     s = (char*)sh+hdrlen;
     fp = ((unsigned char*)s)-1;
     usable = usable-hdrlen-1;
+
+#ifdef USE_JEMALLOC
+    assert(usable <= sdsTypeMaxSize(type));
```
Comment (zhulipeng) [inline src/sds.c]: The assertion here is based on the behavior we expected/monitor from jemalloc, it depends on the internal of jemalloc, is there any guarantee from jemalloc? Maybe a potential risk.

Comment (ranshid) [reply src/sds.c]: why can't we always keep the defending part?  (usable = min(usable, sdsTypeMaxSize(type))
AND also maybe we need to consider keep the assert for the usable size correctness, eg:
assert(usable >= initlen);

Comment (madolson) [reply src/sds.c]:
```
assert(usable <= sdsTypeMaxSize(type));
assert(initlen <= usable);
```
Are our two invariants. Should we be doing those asserts regardless of the allocator being used?

Comment (ranshid) [reply src/sds.c]: Yes that is a good point:
we want to make sure that in case JEMALLOC is used we will SURELY have **real** usable == usable, in order to satisfy #453, so I guess we do have to keep assert(usable <= sdsTypeMaxSize(type)); for JEMALLOC (for example in order to assert when JEMALLOC was compiled with different bin sizes?).
I think (as @madolson suggested) the defensive option is better at this point.

Comment (ranshid) [reply src/sds.c]: I would still change the code so that:

> #ifndef USE_JEMALLOC
> usable = MIN(usable, sdsTypeMaxSize(type));
> #endif
> assert(usable <= sdsTypeMaxSize(type) && initlen <= usable);

Comment (poiuj) [reply src/sds.c]: I think there is a better solution. We can adjust types after the allocation/reallocation. Previously I thought there are some edge cases that wouldn't work. But after reconsidering the idea, and noting again that `alloc` doesn't include header size and null terminator, I think it's a better way.
With this solution we don't need to take assumptions for correctness. Instead of truncating `usable`, we're going to adjust the header type.
I'll update the PR with this idea.

**Comment by poiuj** [comment]
Updated the branch. Includes the new approach with adjusting types. Also rebased on the latest unstable. Unfortunately, missed some style changes during conflict resolution. But it's small stuff that I can fix later before the actual merge.

File: src/sds.c [hunk 6]
```diff
@@ -54,23 +54,27 @@ static inline int sdsHdrSize(char type) {
...
     return SDS_TYPE_32;
 #endif
 }
 
 static inline size_t sdsTypeMaxSize(char type) {
-    if (type == SDS_TYPE_5) return (1 << 5) - 1;
-    if (type == SDS_TYPE_8) return (1 << 8) - 1;
-    if (type == SDS_TYPE_16) return (1 << 16) - 1;
+    if (type == SDS_TYPE_5)
+        return (1<<5) - 1;
```
Comment (ranshid) [inline src/sds.c]: I think it is better to leave the current format in order to avoid non-logical changes.

**Review by ranshid — COMMENTED** [review]
maybe we can make a function to capture the repeated code?:
```
 /* Adjust type if usable won't fit into alloc. */
    if (type != SDS_TYPE_5 && usable > sdsTypeMaxSize(type)) {
        type = sdsReqType(usable);
        hdrlen = sdsHdrSize(type);
        usable = bufsize - hdrlen - 1;
        assert(usable <= sdsTypeMaxSize(type));
    }
```

File: src/sds.c (hunk 6 above)
Comment (lipzhu) [reply src/sds.c:70]: `sdsTypeMaxSize` is used to do a protection before set `alloc` metadata for sds,  the `alloc` meaning here equals actual allocator size - header size - 1. Maybe we should also minus the header size for `sdsTypeMaxSize`?

**Comment by ranshid** [comment]
LGTM but would like to hear @lipzhu

File: src/unit/test_sds.c [hunk 7]
```diff
@@ -262,5 +263,97 @@ int test_sds(int argc, char **argv, int flags) {
...
+    TEST_ASSERT_MESSAGE("len 32 type", (x[-1] & SDS_TYPE_MASK) >= SDS_TYPE_8);
+#ifdef USE_JEMALLOC
+    TEST_ASSERT_MESSAGE("len 32 sdsAllocSize", sdsAllocSize(x) == s_malloc_size(sdsAllocPtr(x)));
+#endif
+    sdsfree(x);
+
+    x = sdsnewlen(NULL, 252);
+    TEST_ASSERT_MESSAGE("len 252 type", (x[-1] & SDS_TYPE_MASK) >= SDS_TYPE_8);
+#ifdef USE_JEMALLOC
+    TEST_ASSERT_MESSAGE("len 252 sdsAllocSize", sdsAllocSize(x) == s_malloc_size(sdsAllocPtr(x)));
```
Comment (zhulipeng) [inline src/unit/test_sds.c:286]: I found the warn message from jemalloc manual. 

> Any discrepancy between the requested allocation size and the size reported by malloc_usable_size() should not be depended on, since such behavior is entirely implementation-dependent.

**Review by zhulipeng — COMMENTED** [review]
LGTM, just one concern.

File: src/unit/test_sds.c (hunk 7 above)
Comment (poiuj) [reply src/unit/test_sds.c:286]: We don't depend on discrepancy between the requested allocation size and the size reported by `malloc_usable_size`. Not in the test, not in the code. The test checks that `sdsAllocSize` returns the same value as `s_malloc_size`. It's always true as `sdsAllocSize` is based on the value we got from `malloc_usable_size` before.
In any case, this test checks that our assumption is correct. As we vendor jemalloc, the only case when it can change (theoretically) is when we update jemalloc.

Comment (zhulipeng) [reply src/unit/test_sds.c:286]: > It's always true as sdsAllocSize is based on the value we got from malloc_usable_size before.

You are right, so we can open assertion not only for jemalloc?

Comment (poiuj) [reply src/unit/test_sds.c:286]: Oh, you're right! This `ifdef` isn't needed anymore. Updated the PR.

**Review by zhulipeng — APPROVED** [review]

**Review by ranshid — APPROVED** [review]

**Review by madolson — APPROVED** [review]
